HW2/P3/mandelbrot.pyx

```diff
@@ -31,16 +44,103 @@ cpdef mandelbrot(np.complex64_t [:, :] in_coords,
+    #                break
+    #            z = z * z + c
+    #        out_counts[i, j] = iter
+
+
+    #with AVX
+    for i in prange(in_coords.shape[0], nogil=True, schedule='static', chunksize = 1, num_threads = 1):
+        #divide y into 8 (start, end, step)
+        for j in range(0,in_coords.shape[1],8):
+
+            #get the real parts of in_coords and store them into cr
+            #we divided all columns into 8 parts and calculated them in parallel
+            c_r = AVX.make_float8((in_coords[i,j+7].real),
+                                 (in_coords[i,j+6].real),
+                                 (in_coords[i,j+5].real),
+                                 (in_coords[i,j+4].real),
+                                 (in_coords[i,j+3].real),
+                                 (in_coords[i,j+2].real),
+                                 (in_coords[i,j+1].real),
+                                 (in_coords[i,j].real))
+
+            #get the imaginary part if in_coords and store them into ci
+            c_i = AVX.make_float8((in_coords[i,j+7].imag),
+                                 (in_coords[i,j+6].imag),
+                                 (in_coords[i,j+5].imag),
+                                 (in_coords[i,j+4].imag),
+                                 (in_coords[i,j+3].imag),
+                                 (in_coords[i,j+2].imag),
+                                 (in_coords[i,j+1].imag),
+                                 (in_coords[i,j].imag))
+
+            #create z for 8 parts, but all 0
+            z_i =  AVX.float_to_float8(0.0)
+            z_r = AVX.float_to_float8(0.0)
+
+            #initialzie the counter for 8 registers
+            counters = AVX.float_to_float8(0.0)
+
+
+
+            #start the iterations
+            for iter in range(max_iterations):
+
+                #check if magnitude of z is less than 4, if not, break the program
+                #comparison of AVX and assign values to z
+                #using mask as comparison
+                #calculate the magnitude of z
+                magz = AVX.add(AVX.mul(z_r,z_r), AVX.mul(z_i,z_i))
+
+                #create a condition(if z is smaller than 4)
+                cond = AVX.less_than(magz, AVX.float_to_float8(4.0))
+
+                #check the condition
+                #if the condition is true, so do the calculation
+                #and the iteration plus one
+                #update the z
+                #if they are all false, then we can break it
+                if(AVX.signs(cond) == 0):
+                    break
+                #else we could update the z value
+                else:
+                    #when calculating z_i, we have to use the original real part
+                    #so I used an temporary variable to store it
+                    z_r_temp = z_r
+                    #z_real = z_real ^ 2 - z_imag ^2 + c_real
+                    z_r = AVX.add(AVX.sub(AVX.mul(z_r,z_r),AVX.mul(z_i,z_i)),c_r)
```

Comment:
Thanks, but as professors mentioned in piazza, some machines don't support fmadd, in order to avoid such case, I chose to use add although it's bit hard to read.